### DISCUSSION TITLE
ESPNOE on ESP32-C3 not working

### DISCUSSION THREAD

**Discussion by BOB63** [micropython/micropython#15455] [ESP32] [answered]
Micropython : ESP32_GENERIC_C3-20240602-v1.23.0  , Board ESP32C3 mini , Thonny  4.1.4.
I'm trying to use espnow on two mini board ESP32-C3 but seems that aren't able to eneter in communication and also changing both board the problem is the same.

As code i'm using the examples found in the doc , adding just few changes .
**For the Trasmitter this is the code :**
`import network
import espnow
import utime
# A WLAN interface must be active to send()/recv()
sta = network.WLAN(network.STA_IF)  # Or network.AP_IF
sta.active(True)

e = espnow.ESPNow()
e.active(True)
peer = b'd\xe83\x8a\xdd\xcc'   # MAC address of peer's wifi interface
e.add_peer(peer)      # Must add_peer() before send()
print('Peer MAC : ', e.get_peers())
utime.sleep_ms(50)
e.send(peer, "Starting...")

for i in range(1000):
    e.send(peer, str(i)*20, True)
e.send(peer, b'end', True)  

`

**For the Receiver this is the code:**

`import network
import espnow
import utime

# A WLAN interface must be active to send()/recv()
sta = network.WLAN(network.STA_IF)
sta.active(True)
#sta.disconnect()   # Because ESP8266 auto-connects to last Access Point
mac_address = sta.config("mac")
print(mac_address)
print("Device MAC Address tx:", ":".join(["{:02X}".format(byte) for byte in mac_address]))

e = espnow.ESPNow()
e.active(True)

while True:
    
    print("wait")
    host, msg = e.recv()
    print("received")
    if msg:             # msg == None if timeout in recv()
        print(host, msg)
        if msg == b'end':
           break
    print('no data received')    `

This is what i see on the Thonny console from the transmitter :

>>> %Run -c $EDITOR_CONTENT

MPY: soft reboot
Peer MAC :  ((b'd\xe83\x8a\xdd\xcc', b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00', 0, 0, False),)

While from the receiver I got :

_>>> %Run -c $EDITOR_CONTENT

MPY: soft reboot
b'd\xe83\x8a\xdd\xcc'
Device MAC Address tx: 64:E8:33:8A:DD:CC
wait_

Seems that the receiver  code still waiting to receive data from the trasmitter.
In the C libraries for ESPnow used with Arduino iDE there were a function to see from the trasmitter side if the data were sent .
Is there somthing similar also in micropython ?

What other test can I do , having already changed also both board ?
tks Roberto

**Comment by robert-hh** [comment]
Which boards do you actually use. There are micro ESP32C3 boards sold at very low price which have a very bad antenna design.

Side note: Use lines with 3 backticks ```  before and after the python code for proper display.

**Comment by BOB63** [comment]
Hi Robert,
the boards are similar to this  https://www.amazon.it/ESP32-C3FN4-Processore-Single-core-Funzionamento-Multifunzione/dp/B0CHYJVZ1V and used with ESPnow library on Arduino ide are working perfectly . 
About the formatting , I've used the icons <> from the bar above the editing area but probably are wrong becaise the place just " '  " at the beginning and at the end. I'll take care next time. 
![immagine](https://github.com/user-attachments/assets/af780d80-391f-4e6b-96b7-d67b43683488)

**Comment by robert-hh** [comment]
These are indeed the boards with the poor antenna design. However, at a short distance of a few meters they should work.

About formatting: When using the <> icons for code, pleas highlight the code before pushing <>. Then the lines with three backticks are added. If you just highlight short text, it adds single backticks. Pushing the <> icon is the as as entering Ctrl-E.

**Comment by shariltumin** [comment]
You can try running both of these scripts.

```python
# sender.py

import network
import espnow

# A WLAN interface must be active to send()/recv()
ap = network.WLAN(network.AP_IF)  # Or network.AP_IF
ap.active(False)

sta = network.WLAN(network.STA_IF)  # Or network.AP_IF
sta.active(True)
sta.disconnect()      # disconnect from access-point (if any)

e = espnow.ESPNow()
e.active(True)
bcst = b'\xff\xff\xff\xff\xff\xff'   # MAC address of peer's wifi interface
e.add_peer(bcst)                     # Must add_peer() before send() on esp32
                                     # esp8266 don't listen to broadcast message

e.send(bcst, "Starting...")
for i in range(10):
    e.send(bcst, str(i)*10, True)
    host, msg = e.recv() # Blocking
    if msg:              # msg == None if timeout in recv()
        print(host, msg)

e.send(bcst, b'end')


```

```python
# receiver.py

import network
import espnow

# run this on esp32-xx
#
# A WLAN interface must be active to send()/recv()
ap = network.WLAN(network.AP_IF)
ap.active(False)

sta = network.WLAN(network.STA_IF)
sta.active(True)
sta.disconnect() # disconnect if MCU auto-connects to last Access Point

e = espnow.ESPNow()
e.active(True)

while True:
    host, msg = e.recv() # Blocking
    if msg:             # msg == None if timeout in recv()
        print(host, msg)
        # host = mac address of sender
        # now we can use the mac address to reply
        try:
           e.add_peer(host)
        except:
           pass
        e.send(host, 'Hello')
        if msg == b'end':
            break
```

These are my run logs. I use mpremote, I think you should do the same.
Start the receiver first.

```
$ mpremote a0 mount .
Local directory . is mounted at /remote
Connected to MicroPython at /dev/ttyACM0
Use Ctrl-] or Ctrl-x to exit this shell
>
MicroPython v1.24.0-preview.10.gc1a6b95bf.dirty on 2024-07-12; ESP32-C3 BlueBoard CUSB (KAKI5-NTNG) with ESP32-C3
>>> import receiver
b'`U\xf9t=\xb0' b'Starting...'
b'`U\xf9t=\xb0' b'0000000000'
b'`U\xf9t=\xb0' b'1111111111'
b'`U\xf9t=\xb0' b'2222222222'
b'`U\xf9t=\xb0' b'3333333333'
b'`U\xf9t=\xb0' b'4444444444'
b'`U\xf9t=\xb0' b'5555555555'
b'`U\xf9t=\xb0' b'6666666666'
b'`U\xf9t=\xb0' b'7777777777'
b'`U\xf9t=\xb0' b'8888888888'
b'`U\xf9t=\xb0' b'9999999999'
b'`U\xf9t=\xb0' b'end'
>>> 
```
Then start the sender.

```
$ mpremote a1 mount .
Local directory . is mounted at /remote
Connected to MicroPython at /dev/ttyACM1
Use Ctrl-] or Ctrl-x to exit this shell
>
MicroPython v1.24.0-preview.10.gc1a6b95bf.dirty on 2024-07-12; ESP32-C3 BlueBoard CUART (KAKI5-NTNG) with ESP32-C3
>>> import sender
b'`U\xf9|.\x10' b'Hello'
b'`U\xf9|.\x10' b'Hello'
b'`U\xf9|.\x10' b'Hello'
b'`U\xf9|.\x10' b'Hello'
b'`U\xf9|.\x10' b'Hello'
b'`U\xf9|.\x10' b'Hello'
b'`U\xf9|.\x10' b'Hello'
b'`U\xf9|.\x10' b'Hello'
b'`U\xf9|.\x10' b'Hello'
b'`U\xf9|.\x10' b'Hello'
>>>
```
I hope these are of some help.

If you are interested in learning more about dynamic meshing with ESPNow, please take a look at this [repository](https://github.com/shariltumin/mesh-espnow-micropython/tree/main).

**Reply by BOB63** [reply]
@shariltumin , 
nothing to do , also you version of program dosen' t work with Thonny.
In the receiver.py ,at the beginning of while loop I've put a print ('wait') to check if the loop was working and " print("read") after  
" host, msg = e.recv() # Blocking" . As result is printed only one time "wait" so I've the doubt that the instruction
 " host, msg = e.recv() " block the remaining code if are not received data,  also if I think that after a certain timeout should pass over . 
This afternoon I'll try also mpremote that I've never used before.

**Reply by shariltumin** [reply]
These should should work with thonny too. I start two thonny instances, one for receiver.py and the other for sender.py.

![Screenshot at 2024-07-13 11-56-17](https://github.com/user-attachments/assets/8e940fa3-419d-48a7-ad1d-a7574355f980)

**Comment by glenn20** [comment]
As suggested by @robert-hh , it is much easier to help if you format your code so we can read it more easily. Follow the advice at https://github.com/orgs/micropython/discussions/9111.

> Micropython : ESP32_GENERIC_C3-20240602-v1.23.0 , Board ESP32C3 mini , Thonny 4.1.4. I'm trying to use espnow on two mini board ESP32-C3 but seems that aren't able to eneter in communication and also changing both board the problem is the same.
> 
> As code i'm using the examples found in the doc , adding just few changes . **For the Trasmitter this is the code :**

```py
import network
import espnow
import utime

# A WLAN interface must be active to send()/recv()
sta = network.WLAN(network.STA_IF) # Or network.AP_IF 
sta.active(True)
e = espnow.ESPNow()
e.active(True)
peer = b'd\xe83\x8a\xdd\xcc' # MAC address of peer's wifi interface
e.add_peer(peer) # Must add_peer() before send()
print('Peer MAC : ', e.get_peers())
utime.sleep_ms(50)
e.send(peer, "Starting...")

for i in range(1000):
    e.send(peer, str(i)*20, True)
    e.send(peer, b'end', True)
```
 
> **For the Receiver this is the code:**

```py
import network
import espnow
import utime


# A WLAN interface must be active to send()/recv()
sta = network.WLAN(network.STA_IF)
sta.active(True)
#sta.disconnect() # Because ESP8266 auto-connects to last Access Point
mac_address = sta.config("mac")
print(mac_address)
print("Device MAC Address tx:", ":".join(["{:02X}".format(byte) for byte in mac_address]))

e = espnow.ESPNow() 
e.active(True)

while True:
    print("wait")
    host, msg = e.recv()
    print("received")
    if msg:             # msg == None if timeout in recv()
        print(host, msg)
        if msg == b'end':
            break
    print('no data received')
```

> 
> 
> This is what i see on the Thonny console from the transmitter :
> 
> > > > %Run -c $EDITOR_CONTENT

```
MPY: soft reboot 
Peer MAC : ((b'd\xe83\x8a\xdd\xcc', b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00', 0, 0, False),)
```
> While from the receiver I got :
> 
> _>>> %Run -c $EDITOR_CONTENT

```
MPY: soft reboot 
b'd\xe83\x8a\xdd\xcc' 
Device MAC Address tx: 64:E8:33:8A:DD:CC 
wait_
```

> Seems that the receiver code still waiting to receive data from the trasmitter. In the C libraries for ESPnow used with Arduino iDE there were a function to see from the trasmitter side if the data were sent . Is there somthing similar also in micropython ?

Yes. Check the documentation `ESPNow.send()` at https://docs.micropython.org/en/latest/library/espnow.html#espnow.ESPNow.send.

`e.send()` will return `True` if it received an acknowledgement from the intended recipient. If there was no response, it will return `False`.

So, you could try printing the return value from `e.send()` to check if the message was received.

> What other test can I do , having already changed also both board ? tks Roberto

One thing to check if you are having problems is to make sure that you perform a hard reset before running your test, as the wifi radio can sometimes be left in a bad state after a soft reset.

**Reply by glenn20** [reply]
Oh - and a little tip: `mac_address.hex()` is a convenient way to print the mac address as a hex string (without the colons though).

**Comment by BOB63** [comment, marked as answer]
At the end as suggeste by @robert-hh the problem is due to the real poor quality of these board. On a set of 8 board ,all bougt from the same supplier ,I've found some board defective . Using the right boards the code suggested by @shariltumin  works fine. What is really strange is that in most of the cases  the boards are ok as trasmitter but fails as used as receiver . Another thing that I've noticed is that the esp32 chip are very hot but I've no way to compare with other kind of esp 32 boards. Unfortunately for my project I need a very small for factor and the ESP32 C3 mini was the right one, but is not easy have boards of good quality.In a previous set of 4 board bough from another supplier I got all boards with  the espr32 c3 cip with inside flash memory, so unusable . Pay attention if you buy these kind of board on Aliexpress. 
Thanks to all for the support.

**Reply by MiguelMLR9** [reply]
Hi, man, I think I had a similar issue with a ESP32-C3 SUPER MINI board, but I'm coding in c++, so the problem was that:
While the transmitter was working 'ok' and the receiver was 'receiving' all the 'messages' (confirmed with Serial.print() messages in the receiver side), the transmitter never received an status of `ESP_NOW_SEND_SUCESS` in the callback function, so I didn't have a way to know if the receiver was receiving the data successfully.
so what I did was, after call `WiFi.mode(WIFI_STA);` \\to set ESP32 as a Wi-Fi Station:
  1. Set the Wifi Channel in both devices (transmitter and receiver) to 6 (because in CHANNEL 1 the receiver just got around 5% of 'messages'. You need to try with different channels), that was crucial to get 100% of messages received in the receiver. Yet the 'status' in the transmitter was not equal to 'ESP_NOW_SEND_SUCCESS'.
  2. On the receiver: increase the `Tx Power` with: `esp_wifi_set_max_tx_power(CONFIG_ESP_PHY_MAX_WIFI_TX_POWER);` (but you can try with lower values and see if works, because when using the 'max' tx power, the esp32 drains more energy). That was the key to always get the confirmation in the transmitter, after that, the status in the transmitter was always equal to `ESP_NOW_SEND_SUCCESS`. I think this step was necessary because the antenna of the device is weak and/or because the default 'tx power' level is set to a lower value from the bootloader|firmware in that specific board (since I check with another esp32-c3 boards and that step was not necessary).
  
 Well, I don't know how to do it in python, but, if you can find a way to do these 2 steps, it should solve your issue.
 
 P.D. I also use an SEEDUINO XIAO ESP32 C3 board (what Is higher quality board) and also gets hot.
